### PR TITLE
Introduce a configurable test case generator.

### PR DESCRIPTION
**Pull request by prideout**
This lets us generate long lists of settings permutations by writing a
simple JSON spec. For example, the following spec would generate six
`Settings` objects, all of which have SSAO enabled. See the unit test
for a larger example.

```
{
    "name": "viewopts",
    "base": { "view.ssao.enabled": true }
    "permute": {
        "view.dithering": ["NONE", "TEMPORAL"],
        "view.sampleCount": [1, 4, 8]
    }
}
```

### CODE REVIEW

File: libs/viewer/include/viewer/Automation.h [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+// Consumes a JSON string and produces a list of automation specs using a counting algorithm.
+//
+// Each top-level item in the JSON is an object with "name", "base" and "permute".
+// The "base" object specifies a single set of changes to apply to default settings.
+// The optional "permute" object specifies a cross product of changes to apply to the base.
+// See the unit test for an example.
+//
+// - Returns true if successful.
+// - This function writes warnings and error messages into the utils log.
+bool generate(const char* jsonChunk, size_t size, std::vector<AutomationSpec>* out);
```
Comment (romainguy) [inline libs/viewer/include/viewer/Automation.h:43]: If this is intended to be a public header (and not just for us), don't use STL

File: libs/viewer/src/parse.h [hunk 2]
```diff
@@ -0,0 +1,33 @@
+/*
```
Comment (romainguy) [inline libs/viewer/src/parse.h:1]: Maybe use a more specific file name?

File: libs/viewer/include/viewer/Automation.h (hunk 1 above)
Comment (prideout) [reply libs/viewer/include/viewer/Automation.h:43]: It's mostly intended for us but let me clean this up anyway and make it into a proper class without any STL in the header, should be easy.

File: libs/viewer/src/parse.h (hunk 2 above)
Comment (prideout) [reply libs/viewer/src/parse.h]: Fixed.

File: libs/viewer/include/viewer/Automation.h (hunk 1 above)
Comment (prideout) [reply libs/viewer/include/viewer/Automation.h]: Fixed.

Comment (pixelflinger) [reply libs/viewer/include/viewer/Automation.h]: I think the "no stl in headers" rule is for libfilament, libutils and libmath -- i.e. what you need to use filament.  I don't think we need to apply it for tools (?)

Comment (romainguy) [reply libs/viewer/include/viewer/Automation.h]: If it's just for us or mostly for us yeah that's fine

File: libs/viewer/src/Automation.cpp [hunk 3]
```diff
@@ -0,0 +1,262 @@
+/*
+ * Copyright (C) 2020 The Android Open Source Project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by mIcable law or agreed to in writing, software
```
Comment (bejado) [inline libs/viewer/src/Automation.cpp:10]: mIcable -> applicable (I think Automation.h, jsonParseUtils.h and SimpleViewer.h have this mutation too)

File: libs/viewer/src/Automation.cpp [hunk 4]
```diff
@@ -0,0 +1,262 @@
...
+            }
+        }
+
+        // Apply changes to the settings object.
+        for (const auto& iter : iters) {
+            const std::string& jsonString = *iter;
+            if (VERBOSE) {
+                slog.i << "  Applying " << jsonString.c_str() << io::endl;
+            }
+            if (!readJson(jsonString.c_str(), jsonString.size(), &base)) {
```
Comment (bejado) [inline libs/viewer/src/Automation.cpp:198]: I assume after this call to `readJson`, we would do something interesting with `base` (like actually testing Filament with it). Am I understanding correctly?

**Review by bejado — APPROVED** [review]
This is really cool

File: libs/viewer/src/Automation.cpp (hunk 4 above)
Comment (prideout) [reply libs/viewer/src/Automation.cpp:198]: Correct, this returns a list of Settings that can then be applied to Filament; my next PR will make use of it.  The current PR merely generates the specified permutations and sets up a unit test for them.

File: libs/viewer/src/Automation.cpp (hunk 3 above)
Comment (prideout) [reply libs/viewer/src/Automation.cpp:10]: Wups, will fix!

**Pull request merged**